### ISSUE TITLE
arduino-esp32 projects fail on macOS 11 "Big Sur"

### ISSUE THREAD

**Issue by adamwilt** [espressif/arduino-esp32#4408]
### Hardware:
Board:							DOIT ESP32 DEVKIT V1
Core Installation version:			esp32 1.0.3
IDE name:						Arduino 1.8.12, 1.8.13
Flash Frequency:					N/A
PSRAM enabled:                                    N/A
Upload Speed:					N/A
Computer OS:                                        macOS 11.0 Beta (20A5384c) (and earlier betas, too)

### Description:
Any attempt to compile an ESP32 project, even an empty one as shown below, fails with:
ValueError: dlsym(RTLD_DEFAULT, kIOMasterPortDefault): symbol not found

The same projects compile and run without problems on macOS 10.5.X.

See also https://forum.arduino.cc/index.php?topic=699989.0 for another report of the same problem.

### Sketch:
```cpp
void setup() {
}

void loop() {
}
```

### Debug Messages:
```
Traceback (most recent call last):
  File "esptool.py", line 57, in <module>
  File "/Library/Python/2.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 389, in load_module
  File "serial/tools/list_ports.py", line 29, in <module>
  File "/Library/Python/2.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 389, in load_module
  File "serial/tools/list_ports_posix.py", line 31, in <module>
  File "/Library/Python/2.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 389, in load_module
  File "serial/tools/list_ports_osx.py", line 32, in <module>
ValueError: dlsym(RTLD_DEFAULT, kIOMasterPortDefault): symbol not found
Failed to execute script esptool
exit status 255
/Applications/Arduino.app/Contents/Java/arduino-builder returned 255
Error compiling for board DOIT ESP32 DEVKIT V1.
```

**Comment by lbernstone** [comment]
https://github.com/pyserial/pyserial/issues/509

**Comment by iwbrhwfy-source** [comment]
me too :(

**Comment by aungthawdar** [comment]
When I compile and upload, 
I got the same error on my BigSir Beta version and solved with the following ways. 

 **-esptool is dependent on pyserial (you need to install pyserial latest version) and** 
 
**-Arduino's default preferences for ESPTool is binary file , it does not use the (esptool.py) source file.**
  So need to replace default esptool binary file with esptool.py file in system.

How to replace?... :)
1. Click Arduino Menu => Preferences
2. Find [ArduinoLibraryPath]/**preferences.txt** at the bottom left of Preferences' Settings Tab,  Click to Open 
    ArduinoLibrary Folder. 
3. Find & Open packages/esp32/hardware/esp32/1.x.x Folder and open platform.txt
4. At the line no 7. replace binary **tools.esptool_py.cmd=esptool** with **tools.esptool_py.cmd=esptool.py** and save.
5. Open packages/esp32/hardware/esp32/1.x.x/tools folder.
6. Copy esptool.py and paste to **ArduinoLibraryFolder/packages/esp32/tools/esptool_py/2.x.x** 
7. Quit Arduino IDE and open again. have a nice coding. 👍 

![Image 11-13-20 at 1 10 AM](https://user-images.githubusercontent.com/1222967/98983204-d79d0200-254e-11eb-8974-fe814c1cc4f0.jpg)

**Comment by robsworld** [comment]
@aungthawdar 👍 

Thank you for this solution. It works great.
The only think I have to add is you must change the permission of esptool.py with the terminal:

chmod +x /Users/XXXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py

**Comment by lbernstone** [comment]
Does the executable work?  If you open a command line and run it, do you get an error (perhaps about python2 not being installed)?  We need a solution that can work on osx and big sur.

**Comment by SpoonS2K** [comment]
@aungthawdar tried your solution and got this error

`Traceback (most recent call last):
  File "/Users/sandybrostin/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py", line 37, in <module>
    import serial
ImportError: No module named serial`

**Comment by me-no-dev** [comment]
can you guys try the idf-release/v4.2 branch? It has updated esptool that should contain a fix

**Comment by TimoWielink** [comment]
@me-no-dev seems to be working with 4.2! Perfect Thnx!

**Comment by adamwilt** [comment]
When I tried replacing esptool with esptool.py in the platform.txt file, I got the `serial ImportError: No module named serial` error. So I tried the 4.2 branch, but I'm clearly missing some steps.  I used the "installation instructions for Mac OS" at https://github.com/espressif/arduino-esp32/blob/idf-release/v4.2/docs/arduino-ide/mac.md, adding the --branch option:
`git clone --branch idf-release/v4.2 https://github.com/espressif/arduino-esp32.git esp32 --depth 1`
Whether I do this on top of my existing installation, or as the initial ESP32 installation after wiping my existing library files, I still get `ValueError: dlsym(RTLD_DEFAULT, kIOMasterPortDefault): symbol not found`. 

What am I doing wrong? What's the proper way to install the idf-release/v4.2 branch?

**Comment by Liferme** [comment]
> When I compile and upload,
> I got the same error on my BigSir Beta version and solved with the following ways.
> 
> **-esptool is dependent on pyserial (you need to install pyserial latest version) and**
> 
> **-Arduino's default preferences for ESPTool is binary file , it does not use the (esptool.py) source file.**
> So need to replace default esptool binary file with esptool.py file in system.
> 
> How to replace?... :)
> 
> 1. Click Arduino Menu => Preferences
> 2. Find [ArduinoLibraryPath]/**preferences.txt** at the bottom left of Preferences' Settings Tab,  Click to Open
>    ArduinoLibrary Folder.
> 3. Find & Open packages/esp32/hardware/esp32/1.x.x Folder and open platform.txt
> 4. At the line no 7. replace binary **tools.esptool_py.cmd=esptool** with **tools.esptool_py.cmd=esptool.py** and save.
> 5. Open packages/esp32/hardware/esp32/1.x.x/tools folder.
> 6. Copy esptool.py and paste to **ArduinoLibraryFolder/packages/esp32/tools/esptool_py/2.x.x**
> 7. Quit Arduino IDE and open again. have a nice coding. 👍
> 
> ![Image 11-13-20 at 1 10 AM](https://user-images.githubusercontent.com/1222967/98983204-d79d0200-254e-11eb-8974-fe814c1cc4f0.jpg)

It really works, and we have to combine the method with the command @robsworld provided. When compiling, remember to choose "ESP32 Arduino", NOT "ESP32 Arduino(in sketch book)", and have fun coding!

**Comment by me-no-dev** [comment]
@TimoWielink thanks for the confirmation :) I have updated the version in Master, so it will go out in 1.0.5 in a few days.

**Comment by aungthawdar** [comment]
> @aungthawdar tried your solution and got this error
> 
> `Traceback (most recent call last): File "/Users/xxxxxxx/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py", line 37, in <module> import serial ImportError: No module named serial`

@SpoonS2K 
You need to install pyserial package and be sure to import first 

**https://pyserial.readthedocs.io/en/latest/pyserial.html**
esptool.py used(import) **serial** library 

Here is how to test, you've already installed **seriak** or not.

1. Open terminal
2. type **python** in that terminal
3. type **import serial** 
4. if no error, you've already installed serial, 
   other wise you'll get 
    "import serial ImportError: No module named serial
<img width="808" alt="Screen Shot 2020-11-15 at 6 18 01 PM" src="https://user-images.githubusercontent.com/1222967/99184131-4eb0e100-276f-11eb-9540-27a3b473982a.png">
"

**Comment by tomtomhoward** [comment]
Thanks guys, just did the above and got it working on Big Sur for ESP32 in Arduino.
I had to update to 4.2, install python serial, and modify the files.

There's also some playing about getting the CP210x driver working for Dev-Kit ESP32 variants, involving adding the Silicon Labs developed ID to kext in OSX.

**Comment by me-no-dev** [comment]
@tomtomhoward if you have updated to 4.2, then you should not have needed to change esptool. It is already a patched one.

**Comment by tomtomhoward** [comment]
I wouldn't trust my take on it, I'm working at the limits of my knowledge here to be honest. I think I might have messed up the 4.2 install as my Documents folder is on iCloud Drive.

**Comment by wick3r** [comment]
> > @aungthawdar tried your solution and got this error
> > `Traceback (most recent call last): File "/Users/xxxxxxx/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py", line 37, in <module> import serial ImportError: No module named serial`
> 
> @SpoonS2K
> You need to install pyserial package and be sure to import first
> 
> **https://pyserial.readthedocs.io/en/latest/pyserial.html**
> esptool.py used(import) **serial** library
> 
> Here is how to test, you've already installed **serial** or not.
> 
> 1. Open terminal
> 2. type **python** in that terminal
> 3. type **import serial**
> 4. if no error, you've already installed serial,
>    other wise you'll get
>    "import serial ImportError: No module named serial
> 
> <img alt="Screen Shot 2020-11-15 at 6 18 01 PM" width="808" src="https://user-images.githubusercontent.com/1222967/99184131-4eb0e100-276f-11eb-9540-27a3b473982a.png">
> 
> Have a nice code. :P

Thank you so much for taking the time to explain! I did all this and got it to work. :)

**Comment by olivierbrcknr** [comment]
Thank all of you! I tried out different variants but still was not able to get it to work.

I reinstalled the Arduino IDE and deleted all my library folders to start from a clean slate. As far as I have understood, the current main branch has already the Big Sur patch included, right? But I unfortunately still get the `ValueError: dlsym(RTLD_DEFAULT, kIOMasterPortDefault): symbol not found` error.

After doing the recommended fix by @aungthawdar I get also the `ImportError: No module named serial` error which I am not able to fix through installing *pyserial* and *serial* in both, python2.7 and python3.8.

Do you have any ideas or recommendations? Would be very much appreciated!

**Comment by me-no-dev** [comment]
@olivierbrcknr did you install from git? to your [sketchbook]/hardware/espressif/esp32 folder?

**Comment by olivierbrcknr** [comment]
Yes, I followed the repo's [instructions for macOS](https://github.com/espressif/arduino-esp32/blob/master/docs/arduino-ide/mac.md)

**Comment by me-no-dev** [comment]
could you check the esptool version?

**Comment by me-no-dev** [comment]
You can now also try to install 1.0.5-rc2 through the board manager using https://raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_dev_index.json

**Comment by olivierbrcknr** [comment]
So, I added the .json file you shared to reinstall it through the board manager. This didn't work, so I deleted and installed everything again to make sure I do not have any old file conflicting or accidentally being used, still the same error. I then figured out I had yet another python version, installed *pyserial* and *serial* there as well to test the *esptool.py* method again, but also no success. 

 I have the esp32 library version `1.0.5-rc2` installed and esptool version `3.0.0` if I see it correctly.

**Comment by me-no-dev** [comment]
hmmm that is strange. that version of esptool I read should be patched... I guess I need to figure out BigSur VM, because I can not now upgrade to it and risk messing up my work env :)

**Comment by olivierbrcknr** [comment]
No worries, I should have done the same ;) I will experiment a little more and try to figure out the issue, probably it's related to my individual settings/mess-ups. I will let you know if I solve it!

**Comment by aungthawdar** [comment]
@olivierbrcknr , 

In my opinion, whether it's latest ArduinoIDE / latest Arduino-esp32 version or not , dependency is python & pyserial.
Your problem is that can't installed **python serial(pyserial)** package correctly for system default using **python** version.

I suggest to **https://opensource.com/article/19/5/python-3-default-mac** for right **python and pip** version.
After set up the right 
1. python ( recommended version is >3.7 , python 2.7 is already deprecated ) and
2. pip ( recommended version is pip3 )

install python -m pip install pyserial 
I think your **ValueError:....** will be gone away soon. 

Have a nice code. :P

**Comment by adamwilt** [comment]
It would be preferable to have a version of the toolchain that _just worked_ with the Python version installed as part of the OS. As it currently stands, installing esp32 library version 1.0.5-rc2 on OS 11 fails exactly the same way as version 1.0.4 does.

Yes, one can install Homebrew, then install Python 3.7.x, then install pip3, and finally make the dotfile changes to run pyenv, but that's hardly a turnkey solution, especially by Arduino standards. And you'd only know about it if you'd followed this issue. Expect a *lot* of support requests if this convoluted path isn't straightened out.  :-) 

For me, it's just a lot less bother to develop in a 10.15 VM atop OS 11, because I'd already gotten that working a couple of weeks ago on the first try, with no false steps and no weird error messages. But that's just me...

**Comment by me-no-dev** [comment]
given that mac uses a prepackaged python into binary, it should not care for what is on your OS. I find it strange that it does. Whatever the fix is, it should not require the user to install anything else, just like it is on anything but BigSur. Arduino does not differentiate between the versions, so we can not provide two files for mac.

**Comment by rOOOObin** [comment]
> @aungthawdar 👍
> 
> Thank you for this solution. It works great.
> The only think I have to add is you must change the permission of esptool.py with the terminal:
> 
> chmod +x /Users/XXXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py



> @aungthawdar 👍
> 
> Thank you for this solution. It works great.
> The only think I have to add is you must change the permission of esptool.py with the terminal:
> 
> chmod +x /Users/XXXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py

I´m not really good in using the terminal on mac.
Can you tell me what to type in at XXXXX ?

Thank you!

**Comment by aungthawdar** [comment]
@rOOOObin , XXXXX means XXX^2 => **porn square** (Just kidding). 
XXXXX means current user's home directory(username) in unix system.

In your mac terminal,
type **chmod +x /Users/** then type the tab key.
You will see the user home directory name.

Cheers

**Comment by me-no-dev** [comment]
Could you please try the esptool binary from the zip? No need to edit platform.txt

[esptool.zip](https://github.com/espressif/arduino-esp32/files/5556528/esptool.zip)

**Comment by adamwilt** [comment]
I tried this after installing 1.0.5-rc2, and replacing esptool with the zipped version. I got _much_ farther: 

I had to first run esptool by right-clicking and selecting Open, then approving it (because it wasn't signed to Apple's liking!).

I then tried a compile & upload:
`Arduino: 1.8.13 (Mac OS X), Board: "DOIT ESP32 DEVKIT V1, 80MHz, 921600, None"

Sketch uses 195784 bytes (14%) of program storage space. Maximum is 1310720 bytes.
Global variables use 13016 bytes (3%) of dynamic memory, leaving 314664 bytes for local variables. Maximum is 327680 bytes.
esptool.py v3.0-dev
Serial port /dev/cu.usbserial-0001
Connecting........_____....._____....._____....._____....._____....._____....._____

A fatal error occurred: Failed to connect to ESP32: Timed out waiting for packet header
A fatal error occurred: Failed to connect to ESP32: Timed out waiting for packet header
`
During this time the ESP32 kept resetting, and when the upload finally failed, the existing code on the ESP32 ran, with proper Arduino serial monitor output, so the serial driver is clearly working as far as basic comms are concerned.

I had the same results at upload speeds of 115200. But if I press and hold the BOOT button on the ESP32 board when the IDE says "Connecting", the upload begins and the code loads and runs properly. Hooray!

I haven't had to press the BOOT button before, but I have to do that now, can live with that.  ;-)

**Comment by me-no-dev** [comment]
hmmm that is not expected behaviour

**Comment by HomeSpan** [comment]
After updating to Big Sur I received the same errors reported here.  Used the board manager to install 1.0.5-rc2 but when trying to compile I now receive the following error:

```
usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file
       cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory
exit status 64
/Applications/Arduino.app/Contents/Java/arduino-builder returned 64
Error compiling for board Adafruit ESP32 Feather.
```
I assume this means I am missing a file somewhere and arduino-builder is trying to copy a non-existent file?

**Comment by me-no-dev** [comment]
it seems like some files are missing. Not an issue with esptool, so you might want to start clean (uninstall any Git version from your sketchbook/hardware folder)

**Comment by ivanisko** [comment]
Hi @me-no-dev,
I try all things that you suggest and still have "Timed out waiting for packet header" message see below.

Thanks for help
Jan

`Sketch uses 957311 bytes (73%) of program storage space. Maximum is 1310720 bytes.
Global variables use 36600 bytes (11%) of dynamic memory, leaving 291080 bytes for local variables. Maximum is 327680 bytes.
esptool.py v3.0-dev
Serial port /dev/cu.usbserial-0001
Connecting........_____....._____....._____....._____....._____....._____....._____

A fatal error occurred: Failed to connect to ESP32: Timed out waiting for packet header
A fatal error occurred: Failed to connect to ESP32: Timed out waiting for packet header`

**Comment by me-no-dev** [comment]
![Screenshot 2020-11-18 at 13 10 52](https://user-images.githubusercontent.com/12663778/99523400-b3528280-299f-11eb-9327-7c95473fc9fd.png)

Works fine for me with FTDI. I was not able to find drivers for SiLabs that work on Big Sur

**Comment by ivanisko** [comment]
Ok, please what are the steps to get it work? Just replace esptool with your link? 
Thanks

**Comment by me-no-dev** [comment]
Yes, that is all I have done. 1.0.5-RC2 (though that should not matter) and replaced the binary with the one from the zip. Connected the ESP through FTDI, selected the port and hit Upload. What adapter are you using? I notice that your serial port has a different name, which means different driver.

**Comment by ivanisko** [comment]
OK, I try other ESP Dev kit and it's working normally after replace binary from zip.
So problem is in ESP.

Thanks

**Comment by lrmate** [comment]
> Could you please try the esptool binary from the zip? No need to edit platform.txt
> 
> [esptool.zip](https://github.com/espressif/arduino-esp32/files/5556528/esptool.zip)

Thanks, this did the job!

**Comment by ardu1** [comment]
Hello, I have also the problem, that I could not compile any sketch for my Wemos Lolin32 after I updated to Big Sur. I tried all things you describe but it will not works. This is the error report:

Traceback (most recent call last):
  File "/Applications/Arduino.app/Contents/Java/portable/packages/esp32/tools/esptool_py/2.6.1/esptool.py", line 37, in <module>
    import serial
ImportError: No module named serial
exit status 1

Does anyone has an idea to fix it?

Thanks and best regards!

**Comment by HomeSpan** [comment]
Yes, after copying esptool.py into into esptool_py/2.6.1/ and editing platform.txt to point to esptool.py (instead of the binary esptool), I received the same error.  To correct, I tried to use pip to install, but pip was not found.  However, I found that my Mac did have pip3.  So from the Terminal, type:

$  pip3 install pyserial

Depending on your permissions, you may need to run this as superuser as follows:

 $  sudo pip3 install pyserial

This seemed to properly install pyserial into the Python3 environment.  I then edited the first line in the esptool.py to use python3 instead of python (which on my Mac defaulted to python 2.7) as follows.  The first line of esptool.py now reads.

#!/usr/bin/env python3

This change to the the first line of esptool.py, as well the pip3 install of pyserial above did the trick for me.  I am able to compile and upload to an Adafruit ESP32 board.

**Comment by ardu1** [comment]
Thanks, I tried it to do and got this answer:
WARNING: The directory '/Users/xxxxx/Library/Caches/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
WARNING: The directory '/Users/ingogrutz/Library/Caches/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
Requirement already satisfied: pyserial in /Library/Python/3.8/site-packages (3.4)
WARNING: You are using pip version 19.2.3, however version 20.2.4 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
What have I to do?

**Comment by HomeSpan** [comment]
Based on your diagnostics, it seems that pyserial is currently installed.  Did you try editing esptool.py and setting the first line to point to python3 instead of python?  When trying to compile do you get the same error (can't import serial)?

**Comment by ardu1** [comment]
Yes I did all the points. Yes, no matter what I try to compile, every time the same error.

**Comment by HomeSpan** [comment]
Maybe try updating pip3 itself per the diagnostic (I received the same suggestion and upgraded before installing pyserial).  Maybe also try running the install with sudo -H as recommended?

**Comment by ardu1** [comment]
Ihm not very fit to use the terminal. Could you please write what commands I have to write?
 Thank you very much!

**Comment by HomeSpan** [comment]
Sure, from the terminal, first try:

$ sudo -H pip3 install --upgrade pip

Then, if that works:

$ pip3 install pyserial

If this second command does not work, try:

$ sudo -H pip3 install pyserial

**Comment by ardu1** [comment]
The installation was running and I have now the version 20.2.4

Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: pyserial in /Library/Python/3.8/site-packages (3.4)

Still the same error :-(

**Comment by HomeSpan** [comment]
Hmm.  To make sure the correct file is being read, try this:  Edit the esptool.py file and add a new second line with some nonsensical command, like GLOP or FOO.   When you try to compile, this should report an error - something like GLOP not found, or unrecognizable command GLOP.  If it does not, that means the build process is reading a different copy of esptool.py file (in other words, not the one you are editing).

The second thing to try is to make sure that pyserial is indeed properly loaded.  From the terminal, enter python3 as follows:

$ python3

This should launch an interactive python session where you can then type at the >>> prompt:

>>> import serial

If this results in an error, that means pyserial is not properly installed.  If it does not result in any output, it suggests the library is indeed installed correctly.  To exit the python, type quit().

**Comment by adamwilt** [comment]
I tried installing the new esptool from esptool.zip on macOS 10.15.7 to see if it behaves any differently. On 10.15.7, it connected to my DOIT ESP32 DEVKIT V1 immediately, just like the 1.0.4 version did. On macOS 11, I have to hold down the BOOT button for it to connect. So it's something in the OS making the difference.... if that's any help.

**Comment by jenspr** [comment]
I am blind or stupid... or both of it at the same time. 
I got that I need to use new esptool binary.. But how do I do that? Where do I need to copy or install it? :-/
Any hint highly appreciated

**Comment by me-no-dev** [comment]
https://github.com/espressif/arduino-esp32/issues/4408#issuecomment-729246012

replace esptool where it is installed. if from git, then in tools/esptool/esptool, else where the package manager puts it.

**Comment by neovo99** [comment]
- `pip uninstall serial`
- `pip install pyserial`

Solved for me. I'm using the default esptool.

**Comment by olivierbrcknr** [comment]
> Could you please try the esptool binary from the zip? No need to edit platform.txt
> 
> [esptool.zip](https://github.com/espressif/arduino-esp32/files/5556528/esptool.zip)

Thank you so much, @me-no-dev! This file solved the issue for me. Sorry for the late reply.

**Comment by migall** [comment]
Thank you olivierbrcknr
esptool.zip
works for me.

The tool was initially blocked by System Security: unknown developer.
Unblock as follows:
    go to System Preferences > Security & Privacy
    Click allow
    go to Arduino IDE > Upload an ESP32 Sketch
    Click Allow at the security prompt

**Comment by jenspr** [comment]
I could replace esptool (binary, from zip linked above).
I found the target folder in 
<username>\library\arduino15\packages\esp32\tools\esptool_py\2.6.1\

I am on macOS BigSur and Apple M1 Silicon.

I also had to unblock. But I did to navigate to the folder. Right click on binary ->open. And confirm the security dialog. 
Afterwards IDE was running the file file.. 

However ran into another problem: **On upload I get "A fatal error occurred: MD5 of file does not match data in flash!"**
now..

**Comment by DeDraak** [comment]
Hello!

I did all I can but now I only have this error:

__Arduino:1.8.13 (Mac OS X), Board:"ESP32 Dev Module, Disabled, Default 4MB with spiffs (1.2MB APP/1.5MB SPIFFS), 240MHz (WiFi/BT), QIO, 80MHz, 4MB (32Mb), 921600, None"

fork/exec /Users/xxxx/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py: permission denied
Fout bij het compileren voor board ESP32 Dev Module__

What is going wrong? How to get the permission?

**Comment by DeDraak** [comment]
> When I compile and upload,
> I got the same error on my BigSir Beta version and solved with the following ways.
> 
> **-esptool is dependent on pyserial (you need to install pyserial latest version) and**
> 
> **-Arduino's default preferences for ESPTool is binary file , it does not use the (esptool.py) source file.**
> So need to replace default esptool binary file with esptool.py file in system.
> 
> How to replace?... :)
> 
> 1. Click Arduino Menu => Preferences
> 2. Find [ArduinoLibraryPath]/**preferences.txt** at the bottom left of Preferences' Settings Tab,  Click to Open
>    ArduinoLibrary Folder.
> 3. Find & Open packages/esp32/hardware/esp32/1.x.x Folder and open platform.txt
> 4. At the line no 7. replace binary **tools.esptool_py.cmd=esptool** with **tools.esptool_py.cmd=esptool.py** and save.
> 5. Open packages/esp32/hardware/esp32/1.x.x/tools folder.
> 6. Copy esptool.py and paste to **ArduinoLibraryFolder/packages/esp32/tools/esptool_py/2.x.x**
> 7. Quit Arduino IDE and open again. have a nice coding. 👍
> 
> ![Image 11-13-20 at 1 10 AM](https://user-images.githubusercontent.com/1222967/98983204-d79d0200-254e-11eb-8974-fe814c1cc4f0.jpg)

After this I get this fault:

fork/exec /Users/xxxx/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py: permission denied
Fout bij het compileren voor board ESP32 Dev Module

How to give this permission?

**Comment by TimoWielink** [comment]
> > When I compile and upload,
> > I got the same error on my BigSir Beta version and solved with the following ways.
> > **-esptool is dependent on pyserial (you need to install pyserial latest version) and**
> > **-Arduino's default preferences for ESPTool is binary file , it does not use the (esptool.py) source file.**
> > So need to replace default esptool binary file with esptool.py file in system.
> > How to replace?... :)
> > 
> > 1. Click Arduino Menu => Preferences
> > 2. Find [ArduinoLibraryPath]/**preferences.txt** at the bottom left of Preferences' Settings Tab,  Click to Open
> >    ArduinoLibrary Folder.
> > 3. Find & Open packages/esp32/hardware/esp32/1.x.x Folder and open platform.txt
> > 4. At the line no 7. replace binary **tools.esptool_py.cmd=esptool** with **tools.esptool_py.cmd=esptool.py** and save.
> > 5. Open packages/esp32/hardware/esp32/1.x.x/tools folder.
> > 6. Copy esptool.py and paste to **ArduinoLibraryFolder/packages/esp32/tools/esptool_py/2.x.x**
> > 7. Quit Arduino IDE and open again. have a nice coding. 👍
> > 
> > ![Image 11-13-20 at 1 10 AM](https://user-images.githubusercontent.com/1222967/98983204-d79d0200-254e-11eb-8974-fe814c1cc4f0.jpg)
> 
> After this I get this fault:
> 
> fork/exec /Users/xxxx/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py: permission denied
> Fout bij het compileren voor board ESP32 Dev Module
> 
> How to give this permission?

chmod +x /Users/XXXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py

**Comment by DeDraak** [comment]
> > > When I compile and upload,
> > > I got the same error on my BigSir Beta version and solved with the following ways.
> > > **-esptool is dependent on pyserial (you need to install pyserial latest version) and**
> > > **-Arduino's default preferences for ESPTool is binary file , it does not use the (esptool.py) source file.**
> > > So need to replace default esptool binary file with esptool.py file in system.
> > > How to replace?... :)
> > > 
> > > 1. Click Arduino Menu => Preferences
> > > 2. Find [ArduinoLibraryPath]/**preferences.txt** at the bottom left of Preferences' Settings Tab,  Click to Open
> > >    ArduinoLibrary Folder.
> > > 3. Find & Open packages/esp32/hardware/esp32/1.x.x Folder and open platform.txt
> > > 4. At the line no 7. replace binary **tools.esptool_py.cmd=esptool** with **tools.esptool_py.cmd=esptool.py** and save.
> > > 5. Open packages/esp32/hardware/esp32/1.x.x/tools folder.
> > > 6. Copy esptool.py and paste to **ArduinoLibraryFolder/packages/esp32/tools/esptool_py/2.x.x**
> > > 7. Quit Arduino IDE and open again. have a nice coding. 👍
> > > 
> > > ![Image 11-13-20 at 1 10 AM](https://user-images.githubusercontent.com/1222967/98983204-d79d0200-254e-11eb-8974-fe814c1cc4f0.jpg)
> > 
> > 
> > After this I get this fault:
> > fork/exec /Users/xxxx/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py: permission denied
> > Fout bij het compileren voor board ESP32 Dev Module
> > How to give this permission?
> 
> chmod +x /Users/XXXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py

I don't understand? 
So far as I think I did the same thing?!

**Comment by DeDraak** [comment]
> @aungthawdar 👍
> 
> Thank you for this solution. It works great.
> The only think I have to add is you must change the permission of esptool.py with the terminal:
> 
> chmod +x /Users/XXXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py

How did you change the permission?

**Comment by fcohen** [comment]
> `Sketch uses 957311 bytes (73%) of program storage space. Maximum is 1310720 bytes.
> Global variables use 36600 bytes (11%) of dynamic memory, leaving 291080 bytes for local variables. Maximum is 327680 bytes.
> esptool.py v3.0-dev
> Serial port /dev/cu.usbserial-0001
> Connecting........_****.....****_....._****.....****_....._****.....****_....._____
> 
> A fatal error occurred: Failed to connect to ESP32: Timed out waiting for packet header
> A fatal error occurred: Failed to connect to ESP32: Timed out waiting for packet header`

I had the same problem. I found my HiLetgo ESP32 developer board requires you to press the Boot button while esptool does its thing. Instructions here show how a 10u capacitor added to the board solves the problem. https://randomnerdtutorials.com/solved-failed-to-connect-to-esp32-timed-out-waiting-for-packet-header/

**Comment by Fabri54** [comment]
Good morning to everybody. 
Before anything else let me state clearly that, here, everybody is well above my knowledge of UNIX, C++ and Python. I have started playing with Arduino and, then, with ESP8266 and ESP-32 based modules approximately 2 years ago and I'm enjoying it. Unfortunately most of the fun has gone away after updating my iMac 27" to Big Sur.
Today I can still compile and flash all Arduino (UNO, NANO, MEGA) as well as ESP8266 NodeMCU but for anything based on ESP32 the Arduino IDE stops just at compilation with the infamous error:
ValueError: dlsym(RTLD_DEFAULT, kIOMasterPortDefault): symbol not found
Failed to execute script esptool
exit status 255.
I've carefully followed to the letter what has been described in the posts above, the last one by DeDraak I believe recaps all the needed steps,  with no results.
The only (? ha!) thing  I don't understand is that in the Arduino folders Python 2.6.1 is mentioned while in the error list:
File "/Library/Python/2.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 389, in load_module
 File "serial/tools/list_ports_osx.py", line 32, in <module>
 version 2.7 is quoted .
 Has anybody any suggestion for me to come out of this irritating mess? The only alternative I have is to BOOTCAMP to W10 but, honestly, it's a bit awkward and, sometimes, unstable. Many thanks in advance

**Comment by g01d10x** [comment]
Greetings fellow nerds !!
I'm using PlatformIO on Mac .. and was very happy to find this today .. as I had been waiting for a while to upgrade to BigSur.
https://www.silabs.com/community/interface/forum.topic.html/vcp_driver_for_macosbigsur110x-SWJK
Just did the install and it works very well .. no issues with PlatformIO Arduino serial monitoring or uploading :)

**Comment by javier1nc** [comment]
> Greetings fellow nerds !!
> I'm using PlatformIO on Mac .. and was very happy to find this today .. as I had been waiting for a while to upgrade to BigSur.
> https://www.silabs.com/community/interface/forum.topic.html/vcp_driver_for_macosbigsur110x-SWJK
> Just did the install and it works very well .. no issues with PlatformIO Arduino serial monitoring or uploading :)

its works 1/2: [https://www.silabs.com/documents/public/software/Mac_OSX_VCP_Driver.zip](https://www.silabs.com/documents/public/software/Mac_OSX_VCP_Driver.zip)

**Comment by adriankae** [comment]
> Yes, that is all I have done. 1.0.5-RC2 (though that should not matter) and replaced the binary with the one from the zip. Connected the ESP through FTDI, selected the port and hit Upload. What adapter are you using? I notice that your serial port has a different name, which means different driver.

So, I followd all steps but I still can't get it to run.
I get this output:

`Traceback (most recent call last):
  File "/Users/XXXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py", line 37, in <module>
    import serial
ImportError: No module named serial
exit status 1
Error compiling for board FireBeetle-ESP32.`

I reinstalled pyserial. I installed serial (not sure if i even need it). I rebooted my mac. I changed the settings file. etc.

Any further suggestions?

**Comment by Fabri54** [comment]
Just to better frame my situation, I'm currently running Arduino 1.8.14 Hourly Built 2020/12/15 11:34
Big Sur 11.1
I faced a few problems while applying the patches (initially with ESP32 and then with ESP8266 and finally with Heltech WiFi-Kit-32) but, today, every board can be flashed (hope these are not the "last famous words...")

There is just one dumb question that comes to my mind, please do not take offence.

Does the FireBeetle-ESP32 board have its own set of board management tools?
In my case the Heltec-WiFi-Kit-32 couldn't count on the Espressif ESP32 standard toolset and I needed to install a specific one with Board Manger under Tools in the Arduino IDE. Perhaps you have to do it

In any case I applied these patches (for ESP32 and ESP8266):

1) copy 

/Users/XXXXX/Library/Arduino15/packages/esp32/hardware/esp32/1.0.4/tools/esptool.py

to 

/Users/XXXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py

2) sudo chmod +x esptool.py 

3) edit the file /Users/XXXXX/Library/Arduino15/packages/esp32/hardware/esp32/1.0.4/platform.txt and change the following line:

tools.esptool_py.cmd=esptool 

into:

tools.esptool_py.cmd=esptool.py

4) save platform.txt, restart ArduinoIDE

And the message:
ValueError: dlsym(RTLD_DEFAULT, kIOMasterPortDefault): symbol not found
disappeared.

If I remember well, I then had a slightly different error using ESP8266 based board and I found another patch:

1)  Open 
~/Library/Arduino15/packages/esp8266/hardware/esp8266/2.7.4/tools/pyserial/serial/tools/list_ports_osx.py

2) Comment out lines 29 and 30 and append these lines:

   iokit = ctypes.cdll.LoadLibrary('/System/Library/Frameworks/IOKit.framework/IOKit')
   cf = ctypes.cdll.LoadLibrary('/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation')

Here is the code to copy and paste in place of lines 29 and 30.:

#iokit = ctypes.cdll.LoadLibrary(ctypes.util.find_library('IOKit'))
#cf = ctypes.cdll.LoadLibrary(ctypes.util.find_library('CoreFoundation'))
iokit = ctypes.cdll.LoadLibrary('/System/Library/Frameworks/IOKit.framework/IOKit')
cf = ctypes.cdll.LoadLibrary('/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation')

At the end of the day, I had applied patches to the directories created for esp32 and Heltec-esp32, as well as for esp8266 , under ~/LibraryArduino15/packages .

Hope this helps

**Comment by cadusk** [comment]
I used to have the same issue and was able to get it working by using python3 (installed via homebrew) and following the steps for manual installation of ESP32 board libraries as described in esp32 documentation:

https://github.com/espressif/arduino-esp32/blob/master/docs/arduino-ide/mac.md

A new version of pyserial will be installed for python3 and it works just fine. 

Also installed BigSur serial drivers downloaded from silabs.com website.

**Comment by indyasys** [comment]
> > Could you please try the esptool binary from the zip? No need to edit platform.txt
> > [esptool.zip](https://github.com/espressif/arduino-esp32/files/5556528/esptool.zip)
> 
> Thank you so much, @me-no-dev! This file solved the issue for me. Sorry for the late reply.

Thank you, this worked for my Big Sur 11.1, just replaced the esptool mentioned above , but it was required to make this file trusted by following terminal command xattr -d com.apple.quarantine /Users/<your user name>/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool

**Comment by vanhasph** [comment]
Any idea why I'm getting this error now? I did all the above but keep getting this error now.
fork/exec /Users/XXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py: no such file or directory
Error compiling for board ESP32 Dev Module.
Thx in advance,
Philippe

**Comment by mrfibreoptic** [comment]
> I could replace esptool (binary, from zip linked above).
> I found the target folder in
> \library\arduino15\packages\esp32\tools\esptool_py\2.6.1\
> 
> I am on macOS BigSur and Apple M1 Silicon.
> 
> I also had to unblock. But I did to navigate to the folder. Right click on binary ->open. And confirm the security dialog.
> Afterwards IDE was running the file file..
> 
> However ran into another problem: **On upload I get "A fatal error occurred: MD5 of file does not match data in flash!"**
> now..
> 
> **EDIT:**
> My bad. I had set the USB programmer Adapter to 3.3V and its wire connected to the 5V Power input of the ESP32-CAM.
> I changed setting to 5V and now everything works fine and as expected.

Hi jenspr

Im also on new apple m1, can you walk me through your entire process please?

thanks

nick

**Comment by vanhasph** [comment]
> Any idea why I'm getting this error now? I did all the above but keep getting this error now.
> fork/exec /Users/XXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py: no such file or directory
> Error compiling for board ESP32 Dev Module.
> Thx in advance,
> Philippe

added the esptool.py file in that directory and now I get an error which I don't get since pyserial is installed.
Pyserial is not installed for /Library/Developer/CommandLineTools/usr/bin/python3. Check the README for installation instructions.
Traceback (most recent call last):
  File "/Users/philippevanhassel/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py", line 38, in <module>
    import serial
ModuleNotFoundError: No module named 'serial'
Multiple libraries were found for "WiFi.h"
 Used: /Users/philippevanhassel/Library/Arduino15/packages/esp32/hardware/esp32/1.0.4/libraries/WiFi
 Not used: /Applications/Arduino.app/Contents/Java/libraries/WiFi_old
Using library WiFi at version 1.0 in folder: /Users/philippevanhassel/Library/Arduino15/packages/esp32/hardware/esp32/1.0.4/libraries/WiFi 
exit status 1
Error compiling for board ESP32 Dev Module.

I've tried for days to find a solution on my own before asking for help but I'm no developer and don't understand why it all doesn't work.

**Comment by sorriso93** [comment]
Hi all, I had the problem with esptool, solved (I think) with the RC6 esp32 distro following the thread. 
But now, only for every esp32 platform and any board in it, when I try to compile on Arduino Ide any sketch compiled correctly before I get this error. I tried to search for a solution but it seems no-one have the same behaviour. I'm working on big sur intel processor osx. 
Probably is not related to esp32, any suggestion?

bash: line 0: [: /Users/yoghibear/Desktop/progr: binary operator expected
usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file
       cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory
exit status 64
/Applications/Sviluppo/Arduino.app/Contents/Java/arduino-builder returned 64
Error compiling for board ESP32 Wrover Module.

**Comment by sorriso93** [comment]
It didn't solved my problem. Doesn't seem related to esptool or pyserial

**Comment by csteamengine** [comment]
> > Yes, that is all I have done. 1.0.5-RC2 (though that should not matter) and replaced the binary with the one from the zip. Connected the ESP through FTDI, selected the port and hit Upload. What adapter are you using? I notice that your serial port has a different name, which means different driver.
> 
> So, I followd all steps but I still can't get it to run.
> I get this output:
> 
> `Traceback (most recent call last): File "/Users/XXXXX/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py", line 37, in <module> import serial ImportError: No module named serial exit status 1 Error compiling for board FireBeetle-ESP32.`
> 
> I reinstalled pyserial. I installed serial (not sure if i even need it). I rebooted my mac. I changed the settings file. etc.
> 
> Any further suggestions?

I have followed all the steps as well and have confirmed that pyserial is installed on my system, but still get this same error. If I run python and type "import pyserial" I get no errors, but the Arduino IDE still gives the error shown above. 

Any updates on this?

**Comment by jtupulis** [comment]
> I used to have the same issue and was able to get it working by using python3 (installed via homebrew) and following the steps for manual installation of ESP32 board libraries as described in esp32 documentation:
> 
> https://github.com/espressif/arduino-esp32/blob/master/docs/arduino-ide/mac.md
> 
> A new version of pyserial will be installed for python3 and it works just fine.

Worked for me too. Thanks for the tip!

PS I had to copy pyserial form Arduino directory to Library.

**Comment by Boyeen** [comment]
> It didn't solved my problem. Doesn't seem related to esptool or pyserial

Me too. I was working until today when a botched library install (Blynk). I uninstalled Arduino (yeah, I know)
and now this problem:

bash: line 0: [: /Users/tim/xxx: binary operator expected
usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file
       cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory
exit status 64
/Applications/Arduino.app/Contents/Java/arduino-builder returned 64
Error compiling for board ESP32 Dev Module.

Any suggestions?

**Comment by luxnaut** [comment]
Yep, still having major issues after doing everything in here....

**Comment by Boyeen** [comment]
Some sketches fail with a bash error as I showed above and others with a "cp"

FWIW  On some scripts the problems go away if there are no comments at the top of the file... weird right?
Put in a comment and they fail to compile, remove it and they compile/run

**Comment by csteamengine** [comment]
For a Mac Big Sur user on a 2015 Macbook Pro, the only solution that worked was to revert my OS to Catalina. Like I mentioned above, none of the suggested solutions worked for me, and I had confirmed PySerial was installed correctly, but not being utilized by Arduino's IDE. 

I had been having many many issues with Big Sur, so I'm actually happy to be back on Catalina. Hope this helps some poor Big Sur Sole fix their issues.

**Comment by luxnaut** [comment]
So, I was able to get it working (apologies if you've already tried this or if I just overlooked it being posted on here). What I found is that Arduino natively uses Python 2.7 instead of the now-standard 3+ that most would have. So what was happening in my case was it was searching for pyserial FOR 2.7 instead. What I had to do was download pyserial, direct terminal to that folder, and then do a manual install for "python2" and now everything works. Note, I was on Big Sur 11.2 and now am on Big Sur 11.3 Beta so this should work for the most recent macOS.

**Comment by ardu1** [comment]
So would it not be much easier to change in the IDE to Phyton 3+, than nobody has to make a workaround?

**Comment by luxnaut** [comment]
> 
> 
> So would it not be much easier to change in the IDE to Phyton 3+, than nobody has to make a workaround?

You would think, but I've not seen this possible anywhere. I believe Arduino themselves has control over this, but I could be wrong.

**Comment by jdgarcia99** [comment]
> So, I was able to get it working (apologies if you've already tried this or if I just overlooked it being posted on here). What I found is that Arduino natively uses Python 2.7 instead of the now-standard 3+ that most would have. So what was happening in my case was it was searching for pyserial FOR 2.7 instead. What I had to do was download pyserial, direct terminal to that folder, and then do a manual install for "python2" and now everything works. Note, I was on Big Sur 11.2 and now am on Big Sur 11.3 Beta so this should work for the most recent macOS.

Could you please detail a little bit how did you do this?
Sorry but I'm still locked into this problem. I'm also in Big Sur 11.2
THX!!!!

**Comment by luxnaut** [comment]
> 
> 
> > So, I was able to get it working (apologies if you've already tried this or if I just overlooked it being posted on here). What I found is that Arduino natively uses Python 2.7 instead of the now-standard 3+ that most would have. So what was happening in my case was it was searching for pyserial FOR 2.7 instead. What I had to do was download pyserial, direct terminal to that folder, and then do a manual install for "python2" and now everything works. Note, I was on Big Sur 11.2 and now am on Big Sur 11.3 Beta so this should work for the most recent macOS.
> 
> Could you please detail a little bit how did you do this?
> Sorry but I'm still locked into this problem. I'm also in Big Sur 11.2
> THX!!!!

Yeah! So you need to download pyserial (https://pypi.org/project/pyserial/) and unzip it. Put it on your desktop, or somewhere easy to navigate to in terminal. Open terminal and direct it to be inside the folder, and then type "sudo python2 setup.py install". This should initiate the download and then all you have to do it go back into Arduino and it should work.

**Comment by jdgarcia99** [comment]
> > > So, I was able to get it working (apologies if you've already tried this or if I just overlooked it being posted on here). What I found is that Arduino natively uses Python 2.7 instead of the now-standard 3+ that most would have. So what was happening in my case was it was searching for pyserial FOR 2.7 instead. What I had to do was download pyserial, direct terminal to that folder, and then do a manual install for "python2" and now everything works. Note, I was on Big Sur 11.2 and now am on Big Sur 11.3 Beta so this should work for the most recent macOS.
> > 
> > 
> > Could you please detail a little bit how did you do this?
> > Sorry but I'm still locked into this problem. I'm also in Big Sur 11.2
> > THX!!!!
> 
> Yeah! So you need to download pyserial (https://pypi.org/project/pyserial/) and unzip it. Put it on your desktop, or somewhere easy to navigate to in terminal. Open terminal and direct it to be inside the folder, and then type "sudo python2 setup.py install". This should initiate the download and then all you have to do it go back into Arduino and it should work.

Thx luxnaut for this

The library installed well. and I restarted the Arduino IDE

Unfortunately, same error saying:

Traceback (most recent call last):
  File "/Users/juande/Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py", line 38, in <module>
    import serial
ModuleNotFoundError: No module named 'serial'
Found many libraries for "WiFi.h"
Used: /Users/juande/Library/Arduino15/packages/esp32/hardware/esp32/1.0.4/libraries/WiFi
 Not used: /Applications/Arduino.app/Contents/Java/libraries/WiFi
exit status 1

**Comment by luxnaut** [comment]
Hmmmm....I genuinely do not know then, I felt lucky to have this work for
me. I'll be sure to update if I find anything more. Sorry!

**Comment by joesoh** [comment]
It works!  Following the instruction and modified file permission. `Library/Arduino15/packages/esp32/tools/esptool_py/2.6.1/esptool.py` to be executable

**Comment by sorriso93** [comment]
> > It didn't solved my problem. Doesn't seem related to esptool or pyserial
> 
> Me too. I was working until today when a botched library install (Blynk). I uninstalled Arduino (yeah, I know)
> and reinstalled.... now this problem:
> 
> **bash: line 0: [: /Users/tim/xxx: binary operator expected usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory exit status 64 /Applications/Arduino.app/Contents/Java/arduino-builder returned 64 Error compiling for board ESP32 Dev Module.**
> 
> Any suggestions?

Just for all big sur users as a future reference. 
I discovered the problem cause for the errore above (bash line 0 ....). The same error appears on Catalina, so it's not a problem only related to Big Sur.
**It is "simply" becouse we have the sketch folder in a iCloud folder, move it locally and magically the error will disappear.** This is due probably to something changed in the compiler or esptool that tries to copy using cp and doesn't accept the path for icloud folders.

**Comment by valepe** [comment]
> > Me too. I was working until today when a botched library install (Blynk). I uninstalled Arduino (yeah, I know)
> > and reinstalled.... now this problem:
> > **bash: line 0: [: /Users/tim/xxx: binary operator expected usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory exit status 64 /Applications/Arduino.app/Contents/Java/arduino-builder returned 64 Error compiling for board ESP32 Dev Module.**
> > Any suggestions?
> 
> Just for all big sur users as a future reference.
> I discovered the problem cause for the errore above (bash line 0 ....). The same error appears on Catalina, so it's not a problem only related to Big Sur.
> **It is "simply" becouse we have the sketch folder in a iCloud folder, move it locally and magically the error will disappear.** This is due probably to something changed in the compiler or esptool that tries to copy using cp and doesn't accept the path for icloud folders.

Same error but my cloud folder is locally synced so it should see a local path. But this path has a space in the middle and the bash error reports only the path until the space. Maybe the error is here?

**Comment by niccolonormani** [comment]
> > > Me too. I was working until today when a botched library install (Blynk). I uninstalled Arduino (yeah, I know)
> > > and reinstalled.... now this problem:
> > > **bash: line 0: [: /Users/tim/xxx: binary operator expected usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory exit status 64 /Applications/Arduino.app/Contents/Java/arduino-builder returned 64 Error compiling for board ESP32 Dev Module.**
> > > Any suggestions?
> > 
> > 
> > Just for all big sur users as a future reference.
> > I discovered the problem cause for the errore above (bash line 0 ....). The same error appears on Catalina, so it's not a problem only related to Big Sur.
> > **It is "simply" becouse we have the sketch folder in a iCloud folder, move it locally and magically the error will disappear.** This is due probably to something changed in the compiler or esptool that tries to copy using cp and doesn't accept the path for icloud folders.
> 
> Same error but my cloud folder is locally synced so it should see a local path. But this path has a space in the middle and the bash error reports only the path until the space. Maybe the error is here?

Thanks valepe for the suggestion, you are right! Just remove any spaces of the folder/sub folders of the sketch's path and it works fine. No further actions needed.

**Comment by valepe** [comment]
> No further actions needed.

My hope is that they fix it anyway. Indeed it's an error.

**Comment by johnty** [comment]
I just ran into this space in the path issue as well (OSX 10.14), leading to the somewhat cryptic "binary operator expected usage". thanks for the notes here! If I recall this didn't happen before but its been a few OSX versions... I just tried to compile the same code from a while back that was stored in a different folder, while the same code checked out to another path (with no spaces) was working fine.

**Comment by jhoughjr** [comment]
several months into this issue and SiLabs are still ignoring it and pretending the driver they released in September of last year works. It does not. 7 months now with horrible support shows me I will never integrate any Silabs products in any design.

**Comment by ace18zz** [comment]
> When I compile and upload,
> I got the same error on my BigSir Beta version and solved with the following ways.
> 
> **-esptool is dependent on pyserial (you need to install pyserial latest version) and**
> 
> **-Arduino's default preferences for ESPTool is binary file , it does not use the (esptool.py) source file.**
> So need to replace default esptool binary file with esptool.py file in system.
> 
> How to replace?... :)
> 
> 1. Click Arduino Menu => Preferences
> 2. Find [ArduinoLibraryPath]/**preferences.txt** at the bottom left of Preferences' Settings Tab,  Click to Open
>    ArduinoLibrary Folder.
> 3. Find & Open packages/esp32/hardware/esp32/1.x.x Folder and open platform.txt
> 4. At the line no 7. replace binary **tools.esptool_py.cmd=esptool** with **tools.esptool_py.cmd=esptool.py** and save.
> 5. Open packages/esp32/hardware/esp32/1.x.x/tools folder.
> 6. Copy esptool.py and paste to **ArduinoLibraryFolder/packages/esp32/tools/esptool_py/2.x.x**
> 7. Quit Arduino IDE and open again. have a nice coding. 👍
> 
> ![Image 11-13-20 at 1 10 AM](https://user-images.githubusercontent.com/1222967/98983204-d79d0200-254e-11eb-8974-fe814c1cc4f0.jpg)

I followed this step to change my documentation, but I got a new error:.

fork/exec /Users/Lerry/Library/Arduino15/packages/esp32/tools/esptool_py/3.0.0/esptool.py: permission denied
Error compiling for board Adafruit ESP32 Feather.


Does anyone know how to solve this one plz?

**Comment by me-no-dev** [comment]
please stop "hacking" :) Latest 1.0.6 release and current master branch both support BigSur. I develop on M1 mac here :)

**Comment by hey-om7** [comment]
> me too :(

same here :(

**Comment by rhl-** [comment]
Hi, I am on Big Sur 11.4 I have a board with a `DOIT ESP32 DEVKIT V1` chip. 

I've install esp32 module v 1.0.6 via this location: https://dl.espressif.com/dl/package_esp32_index.json,http://arduino.esp8266.com/stable/package_esp8266com_index.json 

I've installed the latest Silab 2102x driver from here: https://www.silabs.com/documents/public/software/Mac_OSX_VCP_Driver.zip

This is my upload experience (while holding the button down on my device):

```
Serial port /dev/cu.SLAB_USBtoUART
Connecting........_____....._____....._____....._____....._____....._____.....____An error occurred while uploading the sketch
_

A fatal error occurred: Failed to connect to ESP32: Timed out waiting for packet header

```

**Comment by igor17400** [comment]
I have decided to write a summary for all those reading the post at a later time. 

1) Open your arduino IDE and press `command + ,` to open preferences window.

2) Click the preferences path at the bottom. This should open finder with the path being used by arduino IDE. If for some reason this doesn't work, make use of a manual step. Open the terminal and run `open /Users/<user>/Library/Arduino15/` 

3) Dive into `/Users/<user>/Library/Arduino15/packages/Heltec-esp32/hardware/esp32/0.0.5/tools/esptool.py` and copy `esptool.py` file. 

4.1) Paste the copied file into `/Users/<user>/Library/Arduino15/packages/Heltec-esp32/tools/esptool_py/2.6.1/` 

4.2) Open `/Users/<user>/Library/Arduino15/platform.txt` file and change `tools.esptool_py.cmd=esptool` to `tools.esptool_py.cmd=esptool.py` 

5) Run `ls /Users/<user>/Library/Arduino15/packages/Heltec-esp32/tools/esptool_py/2.6.1/`  to check if the file was indeed copied. You should see two files, `esptool` and  `esptool.py`. However, `esptool` - without `py` extension - isn't nedded. 

6) Quit and reopen arduino IDE. Try to run the script on the IDE.

7) If the script does not work, after all those steps, due to the following error `ImportError: No module named serial` read the next steps.

8) Open your terminal and make sure you aren't in any virtual environment, unless you're running your arduino IDE in some virtual environment. If you aren't, which was my case, I tried several times to run `python -m pip install pyserial`, however I kept receiving the same error `ImportError: No module named serial`. The problem was that `pyserial` library was being installed on python3 environment and my Arduino IDE was running at python2. 

9) Install pip for `python2` via `curl https://bootstrap.pypa.io/pip/2.7/get-pip.py -o get-pip.py` afterwards run `python2 get-pip.py`

10) Run `python2 -m pip install pyserial` and try to run the script again in arduino IDE. 


Those steps worked for me, I hope it helps 

:)
